lib/core/src/chain_swap.rs

```diff
@@ -104,46 +106,56 @@ impl ChainSwapStateHandler {
             current_height
         );
         for swap in chain_swaps {
-            if let Err(e) = self.rescan_chain_swap(&swap, current_height).await {
+            if let Err(e) = self.rescan_incoming_chain_swap(&swap, current_height).await {
                 error!("Error rescanning Chain Swap {}: {e:?}", swap.id);
             }
         }
         Ok(())
     }
 
-    async fn get_confirmed_unspent_sat(&self, swap: &ChainSwap) -> Result<u64> {
-        let swap_script = swap.get_lockup_swap_script()?.as_bitcoin_script()?;
-        let script_pubkey = swap_script
-            .to_address(self.config.network.as_bitcoin_chain())
-            .map_err(|e| anyhow!("Error getting script address: {e:?}"))?
-            .script_pubkey();
-        Ok(self
-            .bitcoin_chain_service
-            .lock()
-            .await
-            .script_list_unspent(script_pubkey.as_script())?
-            .iter()
-            .filter(|u| u.height > 0)
-            .map(|u| u.value)
-            .sum())
-    }
-
-    async fn rescan_chain_swap(&self, swap: &ChainSwap, current_height: u32) -> Result<()> {
+    async fn rescan_incoming_chain_swap(
+        &self,
+        swap: &ChainSwap,
+        current_height: u32,
+    ) -> Result<()> {
         let monitoring_block_height =
             swap.timeout_block_height + CHAIN_SWAP_MONTIORING_PERIOD_BITCOIN_BLOCKS;
         let is_swap_expired = current_height > swap.timeout_block_height;
         let is_monitoring_expired = current_height > monitoring_block_height;
 
-        match swap.state {
-            Pending if swap.refund_tx_id.is_some() => {
-                // There is a refund tx and the state is Pending. If the funds sent to the
-                // lockup script address are spent then set the state back to Complete/Failed.
-                let confirmed_unspent_sat = self.get_confirmed_unspent_sat(swap).await?;
-                if confirmed_unspent_sat == 0 {
-                    let to_state = match swap.claim_tx_id {
-                        Some(_) => Complete,
-                        None => Failed,
-                    };
+        if (is_swap_expired && !is_monitoring_expired) || swap.state == RefundPending {
+            let swap_script = swap.get_lockup_swap_script()?.as_bitcoin_script()?;
+            let script_pubkey = swap_script
+                .to_address(self.config.network.as_bitcoin_chain())
+                .map_err(|e| anyhow!("Error getting script address: {e:?}"))?
+                .script_pubkey();
+            let confirmed_unspent_sat: u64 = self
+                .bitcoin_chain_service
+                .lock()
+                .await
+                .script_list_unspent(script_pubkey.as_script())?
+                .iter()
+                .filter(|u| u.height > 0)
+                .map(|u| u.value)
+                .sum();
+            if confirmed_unspent_sat > 0 && swap.state != Refundable {
```

Comment:
If the swap is in RefundPending state here (and the refund is not confirmed yet) won't this change its state to Refundable ?

Reply:
OK, reading the docs of script_list_unspent it seems it takes into account mempool spends so in fact if the refund is broadcasted (RefundPending state) then we should get zero unspent.
So it seems it will move RefundPending to Failed/Complete ? I think we are not taking into account the mempool spends here.

Reply:
True, fixed